### PR TITLE
[RFC] Allow "spawn" start method for torchinductor workers.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Context: https://github.com/pytorch/pytorch/issues/108586

This PR adds a config to torchinductor such that users can specify the multiprocessing context for TorchInductor workers in codecache.

This would allow users a choice of using "spawn" in multithreaded environments instead of "fork" being hardcoded as the default.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Comment by eellison** [comment]
Reposting comment from the other issue:

Since triton holds the gil on compilation, I wouldn't expect just forking to be any different than setting TORCHINDUCTOR_COMPILE_THREADS to 1.

I have a PR here https://github.com/pytorch/pytorch/pull/108610 and corresponding (WIP) pr here for triton to release gil here: https://github.com/openai/triton/compare/main...eellison:triton:multi_thread?expand=1.

perf run [here](https://hud.pytorch.org/benchmark/torchbench/inductor_no_cudagraphs?startTime=Fri,%2001%20Sep%202023%2015:37:03%20GMT&stopTime=Fri,%2008%20Sep%202023%2015:37:03%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/eellison/527/head&lCommit=57d8219c00d1c0c37a8f78d60e62600dfbc787d2&rBranch=main&rCommit=e787708ad7650bf017f9ed4517619515fc04d257).

It's not all the way there yet.

**Comment by pritamdamania87** [comment]
@eellison Thanks a lot! I'm assuming if that approach works out I should abandon this PR?

**Comment by eellison** [comment]
Yea, I think so. Could you check that `TORCHINDUCTOR_COMPILE_THREADS=1` fixes your problem ?

**Comment by pritamdamania87** [comment]
> Yea, I think so. Could you check that TORCHINDUCTOR_COMPILE_THREADS=1 fixes your problem ?

Haven't had the chance to test with `TORCHINDUCTOR_COMPILE_THREADS=1`, however this patch fixes the problem and using 0 workers also fixes the problem. Also, note that the number of threads here is not really the problem. The problem is with "fork" as I've elaborated here: https://github.com/pytorch/pytorch/issues/108586#issuecomment-1712000346

**Comment by pritamdamania87** [comment]
@eellison Wondering if you have any updates regarding https://github.com/pytorch/pytorch/pull/108610 and https://github.com/openai/triton/compare/main...eellison:triton:multi_thread?expand=1? Can we land this simple fix as we continue to explore that direction?

**Comment by pritamdamania87** [comment]
@eellison Wondering if you have any updates to share?

**Review by eellison — COMMENTED** [review]
Sorry have not made any progress on the alternate approach. @zdevito, any thoughts on this ? seems reasonable to me.

File: test/inductor/test_codecache.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+
+@requires_cuda()
+def test_codecache_spawn():
+    _run_codecache_test("spawn")
+
+
+@requires_cuda()
+def test_codecache_fork():
+    _run_codecache_test("fork")
```
Comment (ezyang) [inline test/inductor/test_codecache.py:39]: This... doesn't actually do anything in CI, does it?

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
needs rebase

**Review by zdevito — APPROVED** [review]
We likely cannot make 'spawn' the default, because it would require every simple script to include `"__main__"` guards, but it is fine to have it as an option to unblock while we figure out a more durable solution.

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -1308,6 +1308,27 @@ def result(self):
...
+# If this process dies abnormally (e.g. segfault)
+# it will not shut down the workers. Instead
+# the workers will have their parent reassigned to the
+# init process. This launches a separate thread to
+# watch for the worker getting reassigned,
+# and cleans it up in this case.
+#
+# This function cannot be an inner function since otherwise mp_context="spawn" would
+# not work for ProcessPoolExecutor since inner functions cannot be pickled.
+def _async_compile_initializer(orig_ppid):
```
Comment (kumpera) [inline torch/_inductor/codecache.py:1320]: We should use safe abstractions for process control like pidfs to avoid race conditions.

Comment (ezyang) [reply torch/_inductor/codecache.py]: preexisting condition

**Comment by ezyang** [comment]
yes I did, you can double check

File: test/inductor/test_codecache.py (hunk 1 above)
Comment (pritamdamania87) [reply test/inductor/test_codecache.py:39]: Do you mean from a test coverage point of view or that the test won't run in CI at all? If you mean the former, it is adding test coverage for "fork" via `torch._inductor.config.worker_start_method`. In the case where we maybe change the default for `worker_start_method` we wouldn't lose test coverage.

**Comment by weifengpy** [comment]
> Reposting comment from the other issue:
> 
> Since triton holds the gil on compilation, I wouldn't expect just forking to be any different than setting TORCHINDUCTOR_COMPILE_THREADS to 1.
> 
> I have a PR here #108610 and corresponding (WIP) pr here for triton to release gil here: https://github.com/openai/triton/compare/main...eellison:triton:multi_thread?expand=1.
> 
> perf run [here](https://hud.pytorch.org/benchmark/torchbench/inductor_no_cudagraphs?startTime=Fri,%2001%20Sep%202023%2015:37:03%20GMT&stopTime=Fri,%2008%20Sep%202023%2015:37:03%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/eellison/527/head&lCommit=57d8219c00d1c0c37a8f78d60e62600dfbc787d2&rBranch=main&rCommit=e787708ad7650bf017f9ed4517619515fc04d257).
> 
> It's not all the way there yet.

I was running torch.compile in multi-thread process group. ``TORCHINDUCTOR_COMPILE_THREADS=1`` is still the solution. Tried ``torch._inductor.config.worker_start_method = spawn`` but could not resolve my problem

```
  File "/home/weif/local/pytorch-official/pytorch/torch/_inductor/compile_fx.py", line 461, in compile_fx_inner
    compiled_graph = fx_codegen_and_compile(
  File "/home/weif/local/pytorch-official/pytorch/torch/_inductor/compile_fx.py", line 743, in fx_codegen_and_compile
    compiled_fn = graph.compile_to_fn()
  File "/home/weif/local/pytorch-official/pytorch/torch/_inductor/graph.py", line 1415, in compile_to_fn
    return self.compile_to_module().call
  File "/home/weif/local/pytorch-official/pytorch/torch/_dynamo/utils.py", line 265, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/weif/local/pytorch-official/pytorch/torch/_inductor/graph.py", line 1362, in compile_to_module
    mod = PyCodeCache.load_by_key_path(
  File "/home/weif/local/pytorch-official/pytorch/torch/_inductor/codecache.py", line 2246, in load_by_key_path
    exec(code, mod.__dict__, mod.__dict__)
  File "/tmp/torchinductor_weif/t6/ct6lwk4xy72tefid7cag4z6s4kznln6arhjt2l5lnv5gs7lc52oa.py", line 75, in <module>
    async_compile.wait(globals())
  File "/home/weif/local/pytorch-official/pytorch/torch/_inductor/codecache.py", line 2808, in wait
    scope[key] = result.result()
  File "/home/weif/local/pytorch-official/pytorch/torch/_inductor/codecache.py", line 2627, in result
    self.future.result()
  File "/home/weif/local/miniconda3/envs/pytorch-3.10/lib/python3.10/concurrent/futures/_base.py", line 458, in result
    return self.__get_result()
  File "/home/weif/local/miniconda3/envs/pytorch-3.10/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```

**Comment by ppwwyyxx** [comment]
https://github.com/pytorch/pytorch/issues/126311 -- it seems the "spawn" start method doesn't work at all.

The unittest in this PR seems to not test 'spawn' at all -- when `AsyncCompile` is imported the pool was already started in `fork` mode.

**Pull request closed without merging**